### ISSUE TITLE
Allow Multiple Mailers

### ISSUE THREAD

**Issue by myassine2** [quarkusio/quarkus#18452]
## Description

Enhance the mailer package to create multiple mailer properties which can be separated by custom key name. In case a service has to use different SMTP servers from different domains.

## Implementation ideas

It can be done through the properties by defining the * as illustrated bellow:
`quarkus.mailer.*.from=YOUREMAIL@gmail.com`

It can be instantiated as follows in the Java code
`@Inject @Named("myCustomerSmtp") MailClient client;`

**Comment by cescoffier** [comment]
Probably not @Named but @Identifier but yes, that could be useful.

**Comment by gsmet** [comment]
I think we should take a step back and try to be more consistent because some parts of Quarkus use `@Named` IIRC, some specific qualifiers and now this `@Identifier` one.

**Comment by cescoffier** [comment]
I agree, we moved from `@Named` to `@Identifier` following the advice from @mkouba. But not everything has migrated. We should probably look and handle both for a bit.

**Comment by myassine2** [comment]
Interesting I did not know you moved to `@identifier`. Can you share a documentation for further reading please.

**Comment by cescoffier** [comment]
The Javadoc of the class mentions why it's "better" than `@Named`: https://javadoc.io/doc/io.smallrye.common/smallrye-common-annotation/latest/io/smallrye/common/annotation/Identifier.html

**Comment by myassine2** [comment]
@cescoffier thank you for the link! I appreciate the support as always.

**Comment by aaronaxisa-ecabs** [comment]
I need this as well. 

Just a thing to keep in mind. I would ideally have the option that these can be dynamically created at runtime not just through application.yml .

For example we have a email service that through REST lets you setup a new email to use for specific types. As such properties would still make it not work properly.

**Comment by geoand** [comment]
This should be fairly easy to do, do we want to add it?

**Comment by geoand** [comment]
@cescoffier WDYT about sneaking this into Quarkus 3? Should be doable :)

**Comment by gsmet** [comment]
I started working on that a while ago and have an in-progress branch.
It's not as easy as it looks though.

I thought I assigned the issue to me so I wonder if there is a duplicate, I will have a look.

**Comment by cescoffier** [comment]
Do you remember what kind of difficulty you encountered?

**Comment by gsmet** [comment]
Closing as a duplicate of https://github.com/quarkusio/quarkus/issues/9064 . Please track this issue instead if you're interested by the progress.

**Comment by geoand** [comment]
Ah, okay

**Comment by geoand** [comment]
Okay. Let me know if you want me to have a look

**Comment by geoand** [comment]
Didn't we face the sample problem in the mongo extension?

**Comment by gsmet** [comment]
The problem is very common.
Every time you have a Map of components in config with only runtime properties, it becomes a problem if you need the keys at build time.

I know we poked @radcortez several times to know if it would be possible for some maps to have an annotation to say that the structure of the keys is fixed at build time and then it would be possible to get the keys at build time.

But that might not be easy :).

**Comment by geoand** [comment]
My point is that for the case of Mongo I _think_ we rely on the injection points in order to force the creation of the proper beans.

**Comment by geoand** [comment]
Yes, it's a tradeoff

**Comment by cescoffier** [comment]
In Redis we had the same issue.

**Comment by geoand** [comment]
I think we should go ahead with the solution we have already used

**Comment by radcortez** [comment]
> The problem is very common. Every time you have a Map of components in config with only runtime properties, it becomes a problem if you need the keys at build time.
> 
> I know we poked @radcortez several times to know if it would be possible for some maps to have an annotation to say that the structure of the keys is fixed at build time and then it would be possible to get the keys at build time.
> 
> But that might not be easy :).

The issue here is when you have sources that are only available at runtime (for instance Vault, Consul, etc), that could contain a Map configuration structure. How do we get them at build time? We can't :(

That is why we force another required structure at build time configuration so we can get the keys that must be available both at build time and runtime.

We could implement the desired behavior. We just need to be aware that keys provided by runtime sources may have no effect (they still override the existing configuration found at build time, but if they provide a new key structure then we will ignore it because it was not present at build time).
